state/models.go

```diff
@@ -734,3 +737,157 @@ func (w *Record) Equal(other Record) bool {
+	return &t.ExecutionRequestCommon
+}
+
+func (t TemplateExecutionRequest) GetExecutionRequestCustom() *map[string]interface{} {
+	return &map[string]interface{}{
+		TemplatePayloadKey: t.TemplatePayload,
+	}
+}
+
+type TemplateJSONSchema map[string]interface{}
+
+type Template struct {
+	TemplateID      string             `json:"template_id"`
+	TemplateName    string             `json:"template_name"`
+	Version         int64              `json:"version"`
+	Schema          TemplateJSONSchema `json:"schema"`
+	CommandTemplate string             `json:"command_template"`
+	ExecutableResources
+}
+
+type CreateTemplateRequest struct {
+	TemplateName    string             `json:"template_name"`
+	Schema          TemplateJSONSchema `json:"schema"`
+	CommandTemplate string             `json:"command_template"`
+	ExecutableResources
+}
+
+type CreateTemplateResponse struct {
+	DidCreate bool     `json:"did_create"`
+	Template  Template `json:"template,omitempty"`
+}
+
+type TemplateUpdateRequest struct {
+	Schema          string `json:"schema"`
+	CommandTemplate string `json:"command_template"`
+	ExecutableResources
+}
+
+func (t Template) GetExecutableID() *string {
+	return &t.TemplateID
+}
+
+func (t Template) GetExecutableType() *ExecutableType {
+	et := ExecutableTypeTemplate
+	return &et
+}
+
+func (t Template) GetExecutableResources() *ExecutableResources {
+	return &t.ExecutableResources
+}
+
+func (t Template) GetExecutableCommand(req ExecutionRequest) (string, error) {
+	var (
+		err    error
+		result bytes.Buffer
+	)
+
+	// Get the request's custom fields.
+	customFields := *req.GetExecutionRequestCustom()
+	executionPayload := customFields[TemplatePayloadKey]
+	if executionPayload == nil {
+		return "", err
+	}
+
+	schemaLoader := gojsonschema.NewGoLoader(t.Schema)
+	documentLoader := gojsonschema.NewGoLoader(executionPayload)
+
+	// Perform JSON schema validation to ensure that the request's template
+	// payload conforms to the template's JSON schema.
+	if _, err = gojsonschema.Validate(schemaLoader, documentLoader); err != nil {
+		return "", err
+	}
+
+	// Create a new template string based on the template.Template.
+	textTemplate, err := template.New("command").Funcs(sprig.TxtFuncMap()).Parse(t.CommandTemplate)
+	if err != nil {
+		return "", err
+	}
+
+	// Dump payload into the template string.
+	if err = textTemplate.Execute(&result, executionPayload); err != nil {
+		return "", err
+	}
+
+	return result.String(), nil
+}
+
+func (t Template) GetExecutableResourceName() string {
+	return t.TemplateID
+}
+
+// NewTemplateID returns a new uuid for a Template
+func NewTemplateID(t Template) (string, error) {
+	uuid4, err := newUUIDv4()
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("tpl-%s", uuid4), nil
+}
+
+func (t *Template) IsValid() (bool, []string) {
+	conditions := []validationCondition{
+		{len(t.TemplateName) == 0, "string [template_name] must be specified"},
```

Comment:
Do we want to enforce any particular convention. For instance `PythonTask` or `python_task`